internal/Yesod/Session/Memcache/Expiration.hs

```diff
@@ -0,0 +1,87 @@
+module Yesod.Session.Memcache.Expiration
+  ( MemcacheExpiration (..)
+  , getCacheExpiration
+  , noExpiration
+  , minDiffTime
+  ) where
+
+import Internal.Prelude
+
+import Database.Memcache.Types qualified as Memcache
+import Session.Timing.Timeout (Timeout (..))
+import Time (NominalDiffTime, UTCTime, utcTimeToPOSIXSeconds)
+
+newtype Exceptions = InvalidExpiration NominalDiffTime
+  deriving stock (Eq, Show)
+  deriving anyclass (Exception)
+
+data MemcacheExpiration
+  = -- | Do not set expiration times; memache will only evict when the cache is full
+    NoMemcacheExpiration
+  | -- | Sessions will be stored in memcache with the same expiration time that we
+    -- send to the HTTP client, the lesser of the idle and absolute timeouts.
+    UseMemcacheExpiration
+
+-- | Get expiration time for Memcache.
+--
+-- This function will throw if an expiration time is out of bounds, but since
+-- this function is called at the time of app initialization, it is acceptable to
+-- throw.
+getCacheExpiration
+  :: (MonadIO m, MonadThrow m)
+  => MemcacheExpiration
+  -> m UTCTime
+  -> Timeout NominalDiffTime
+  -> m Memcache.Expiration
+getCacheExpiration NoMemcacheExpiration _clock _timeout = pure noExpiration
+getCacheExpiration UseMemcacheExpiration clock timeout =
+  case getMinimumTimeout timeout of
+    Nothing -> pure noExpiration
+    Just duration -> do
+      now <- clock
+
+      let timeStamp = utcTimeToPOSIXSeconds now + duration
+
+      unless (expirationWithinBounds timeStamp)
+        $ throwWithCallStack
+        $ InvalidExpiration timeStamp
+
+      pure $ floor timeStamp
+
+-- | Do not expire the session via Memcache's expiration mechanism.
+--
+--  Memcache will evict the session when the cache is full.
+noExpiration :: Memcache.Expiration
+noExpiration = 0
+
+-- | Get minimum of idle and absolute timeouts.
+--
+-- If both are 'Nothing', return 'Nothing'.
+getMinimumTimeout :: Timeout NominalDiffTime -> Maybe NominalDiffTime
+getMinimumTimeout (Timeout idle absolute) = fmap getMin $ idleMin <> absoluteMin
+ where
+  idleMin = Min <$> idle
+  absoluteMin = Min <$> absolute
```

Comment:
This seems somewhat redundant to - and somewhat inconsistent with - the calculation that `Session.Timing.Math.nextExpires` does. The absolute timeout needs to be added to the session creation date, not to the current time.

Reply:
Changed to using the `Session.Timing.Math` module.